### PR TITLE
bugfix: cdn cannot re-download uncompleted task with header param

### PR DESCRIPTION
**Pull request by lowzj**
Signed-off-by: lowzj <zj3142063@gmail.com>

### Ⅱ. Does this pull request fix one issue?

fixes: #1380

### CODE REVIEW

File: supernode/httpclient/origin_http_client.go [hunk 1]
```diff
@@ -132,18 +132,16 @@ func (client *OriginClient) GetContentLength(url string, headers map[string]stri
...
 // IsSupportRange checks if the source url support partial requests.
 func (client *OriginClient) IsSupportRange(url string, headers map[string]string) (bool, error) {
-	// set headers
-	if headers == nil {
-		headers = make(map[string]string)
-	}
-	headers["Range"] = "bytes=0-0"
+	// set headers: headers is a reference to map, should not change it
+	copied := CopyHeader(nil, headers)
+	copied["Range"] = "bytes=0-0"
```
Comment (zhouhaibing089) [inline supernode/httpclient/origin_http_client.go:137]: Is this same as `delete(copied, "Range")`?

**Comment by zhouhaibing089** [comment]
I do have a general question. Header is usually typed with `map[string][]string`.

https://github.com/golang/go/blob/fca286bed3ed0e12336532cc711875ae5b3cb02a/src/net/http/header.go#L21

And it looks like there is a builtin [Clone][2] in standard library.

[2]: https://github.com/golang/go/blob/fca286bed3ed0e12336532cc711875ae5b3cb02a/src/net/http/header.go#L90

File: supernode/httpclient/origin_http_client.go (hunk 1 above)
Comment (lowzj) [reply supernode/httpclient/origin_http_client.go:137]: Yeah, but I'm afraid that the `header` maybe read in other places, so I just copy it.

Comment (lowzj) [reply supernode/httpclient/origin_http_client.go:137]: > Is this same as `delete(copied, "Range")`?

No, it just read 1 byte from the remote server to test whether the server supports `range` request.

Comment (zhouhaibing089) [reply supernode/httpclient/origin_http_client.go:137]: Oh, I see. Thanks!

**Comment by lowzj** [comment]
> I do have a general question. [Header](https://github.com/golang/go/blob/fca286bed3ed0e12336532cc711875ae5b3cb02a/src/net/http/header.go#L21) is usually typed with `map[string][]string`.
> 
> And it looks like there is a builtin [Clone](https://github.com/golang/go/blob/fca286bed3ed0e12336532cc711875ae5b3cb02a/src/net/http/header.go#L90) in standard library.

Yes, the arg `headers` will be parsed into `http.Header` in the `supernode/httpclient`. I think that it's better to change the content after parsing.

**Comment by lowzj** [comment]
> And it looks like there is a builtin [Clone](https://github.com/golang/go/blob/fca286bed3ed0e12336532cc711875ae5b3cb02a/src/net/http/header.go#L90) in standard library.

The `Clone` method is not public until go version 1.13.

**Comment by zhouhaibing089** [comment]
> The Clone method is not public until go version 1.13.

Oh, good to know. If we're using go1.13 for compilation, then I think it is nice to align with it. Otherwise, I think we could just copy it over.

**Comment by Starnop** [comment]
LGTM.

**Pull request merged**